### PR TITLE
Nested tuples (draft)

### PR DESCRIPTION
**Pull request by tristanmenzel**
## Proposed Changes

-
-
-

### CODE REVIEW

File: src/puya/ir/builder/_utils.py [hunk 1]
```diff
@@ -73,13 +73,11 @@ def assign(
...
-            except ValueError as ex:
-                raise InternalError(
-                    "Incompatible multi-assignment lengths", source_location
-                ) from ex
-            names = [(format_tuple_index(name, idx), var_loc) for idx, _ in enumerate(atypes)]
+            # try:
+            #     ((name, var_loc),) = names
+            # except ValueError as ex:
+            raise InternalError("Incompatible multi-assignment lengths", source_location)
+            # names = [(format_tuple_index(name, idx), var_loc) for idx, _ in enumerate(atypes)]
```
Comment (tristanmenzel) [inline src/puya/ir/builder/_utils.py:80]: I shifted the expansion of tuple vars to `handle_assignment` where we still have wtypes which can be used to infer tuple item names

File: src/puya/ir/builder/assignment.py [hunk 2]
```diff
@@ -75,26 +98,43 @@ def _handle_assignment(
...
                 raise CodeError("unpacking vs result length mismatch", assignment_location)
-            return [
-                val
-                for dst, src in zip(items, source, strict=True)
-                for val in handle_assignment(
-                    context, target=dst, value=src, assignment_location=assignment_location
+
+            results = list[Value]()
+            offset = 0
+            for item in items:
```
Comment (tristanmenzel) [inline src/puya/ir/builder/assignment.py:121]: I'm sure there's a nicer way to do this, but the goal is to match the arity of a assignment target to a slice of the source values (eg. tuple of len == 3 will consume 3 values)

File: src/puya/ir/builder/main.py [hunk 3]
```diff
@@ -401,25 +401,41 @@ def visit_checked_maybe(self, expr: awst_nodes.CheckedMaybe) -> TExpression:
         )
         return value
 
+    def _expand_tuple_var(
+        self, name: str, wtype: wtypes.WTuple, *, source_location: SourceLocation
+    ) -> ValueTuple:
+        return ValueTuple(
```
Comment (tristanmenzel) [inline src/puya/ir/builder/main.py:407]: Flatten nested tuples into a single value tuple

File: src/puya/ir/builder/main.py [hunk 4]
```diff
@@ -504,11 +520,17 @@ def visit_method_constant(self, expr: puya.awst.nodes.MethodConstant) -> TExpres
...
     def visit_tuple_expression(self, expr: awst_nodes.TupleExpression) -> TExpression:
-        items = []
+        items = list[Value]()
         for item in expr.items:
-            # TODO: don't rely on a pure function's side effects (raising) for validation
-            wtype_to_ir_type(item)
-            items.append(self.visit_and_materialise_single(item))
+            if isinstance(item.wtype, wtypes.WTuple):
+                nested = item.accept(self)
+                assert isinstance(nested, ValueTuple)
```
Comment (achidlow) [inline src/puya/ir/builder/main.py:527]: how are you planning to handle other multi-valued types? such as user subroutines or intrinsics that return multiple values?

**Comment by achidlow** [comment]
Approach is looking good, let's just make sure we have sufficient test coverage and also go through any sites in the IR layers where there's potentially existing assumptions about tuples not being nested.

**Comment by tristanmenzel** [comment]
Closing in favour of https://github.com/algorandfoundation/puya/pull/286

**Pull request closed without merging**